src/utils/hd-key.ts

```diff
@@ -56,4 +59,69 @@ export class HdKey {
 
     return derivedDescendantPublicKey;
   }
+
+  /**
+   * Derives a hardened hierarchical deterministic private key.
+   */
+  public static async derivePrivateKeyBytes(privateKey: Uint8Array, relativePath: string[]): Promise<Uint8Array> {
+    HdKey.validateKeyDerivationPath(relativePath);
+
+    let currentPrivateKey = privateKey;
+    for (const segment of relativePath) {
+      const segmentBytes = Encoder.stringToBytes(segment);
+      currentPrivateKey = await HdKey.deriveKeyUsingHkdf({
+        hashAlgorithm      : 'SHA-256',
+        initialKeyMaterial : currentPrivateKey,
+        info               : segmentBytes, // use the segment as the application specific info for key derivation
+        keyLengthInBytes   : 32 // 32 bytes = 256 bits
+      });
+    }
+
+    return currentPrivateKey;
+  }
+
+  /**
+   * Derives a key using  HMAC-based Extract-and-Expand Key Derivation Function (HKDF) as defined in RFC 5869.
+   * TODO: Consolidate HKDF implementation and usage with web5-js - https://github.com/TBD54566975/dwn-sdk-js/issues/742
+   */
+  public static async deriveKeyUsingHkdf(params: {
+    hashAlgorithm: 'SHA-256' | 'SHA-384' | 'SHA-512',
+    initialKeyMaterial: Uint8Array,
+    info: Uint8Array,
+    keyLengthInBytes: number
+  }): Promise<Uint8Array> {
+    const { hashAlgorithm, initialKeyMaterial, info, keyLengthInBytes } = params;
+
+    const webCrypto = getWebcryptoSubtle() as SubtleCrypto;
+
+    // Import the `initialKeyMaterial` into the Web Crypto API to use for the key derivation operation.
+    const webCryptoKey = await webCrypto.importKey('raw', initialKeyMaterial, { name: 'HKDF' }, false, ['deriveBits']);
+
+    // Derive the bytes using the Web Crypto API.
+    const derivedKeyBuffer = await crypto.subtle.deriveBits(
+      {
+        name : 'HKDF',
+        hash : hashAlgorithm,
+        salt : new Uint8Array(0), // `info` should be sufficient in our use case
+        info
+      },
+      webCryptoKey,
+      keyLengthInBytes * 8 // convert from bytes to bits
+    );
+
+    // Convert from ArrayBuffer to Uint8Array.
+    const derivedKeyBytes = new Uint8Array(derivedKeyBuffer);
+    return derivedKeyBytes;
+  }
+
+  /**
+   * Parses the given key derivation path.
+   * @returns Path segments if successfully validate the derivation path.
+   * @throws {DwnError} with `DwnErrorCode.HdKeyDerivationPathInvalid` if derivation path fails validation.
```

Comment:
We aren't returning any path segments here. Since it's only checking if there are empty strings, figured it would be good to put in the comments.

```suggestion
   * Validates that no empty strings exist within the derivation path segments array.
   * @throws {DwnError} with `DwnErrorCode.HdKeyDerivationPathInvalid` if derivation path fails validation.
```

Reply:
Thanks! Keen eyes as always!